### ISSUE TITLE
実験結果について

### ISSUE THREAD

**Issue by guisesueyuk** [dbgroup-nagoya-u/202304-bachelor-thesis-sugie#4]
@baycedar 

次元数の変化による範囲検索で試しにグラフを作ったところ，uart変更前のスループットが1000万を切ってr-boostよりも低性能な結果になってしまいましたが原因がわかりません．正常な範囲ですか？

また，スループットの結果は，同じ条件でREPEAT＿COUNTの数だけ繰り返されるかため平均値を使うであっていますか？

**Comment by baycedar** [comment]
> uart変更前のスループットが1000万を切ってr-boostよりも低性能な結果になってしまいましたが原因がわかりません．正常な範囲ですか？

さすがに見ないとなんとも言えないので、グラフの図を貼ってもらうか、追加したものをpushしてもらえますか？それはそれとして、変更前のUARTは挿入するレコードの分布によって性能が大きく変わる（鈴木君の修論にあったような分布の偏りだけでなく、挿入するレコード数によっても性能が変わる）ことが予想されるので、大負けしたり逆に勝ったりは普通にあると思います。その点が改善後のUARTで解消され、性能が安定化したかどうかも今回のポイントになります。

> スループットの結果は，同じ条件でREPEAT＿COUNTの数だけ繰り返されるかため平均値を使うであっていますか？

はい、そうです。どこを起点に範囲操作が行われるか（および、シミュレーションの場合は生成されるレコードも）は実行するたびに変わるので、複数回測ってその平均値で性能を評価してください。

**Comment by baycedar** [comment]
なお、実験結果はExcelなどでまとめてもよいですし、Pythonでpandasなどを使ってもよいです。自分はpandasで集約結果をcsv（ないしExcelファイル化）してからPowerPointで論文用のPDFにしていますが、matplotlibなどでキレイに作れれば全部Pythonでやってもらってもよいです（自分はまだ使いこなせず痒いところに手が届かないので、PowerPointでやってます）。giantsの `/home/sugiura/workspace/python/csv_to_graph` に変換に使ったノートブックがあるので、必要であれば参照してください。

**Comment by guisesueyuk** [comment]
![画像1](https://github.com/dbgroup-nagoya-u/202304-bachelor-thesis-sugie/assets/129254292/7498288c-4207-40da-b38d-356d39917281)
これです．
アドバイスもありがとうございます．

**Comment by baycedar** [comment]
はい、（skew、レコード数などはわかりませんが）このくらいの結果なら普通に出ると思います。改善後がだいたい2～3倍くらいになるみたいなので、読み取りならR-treeと同等以上、書き込みならUB木と同等みたいな結果になるんじゃないかと思います。

**Comment by guisesueyuk** [comment]
@baycedar 
scanのグラフについてです．

横軸が次元，データ数，選択率，スキューのグラフ全部用意しようと思います．
スキューは固定してそれ以外のパラメータは最大と最小の2つ用意しようと思います．
次元，データ数，選択率すべてが影響していてスキューだけは全く影響がないように見えるためです．
間に合うかはわかりませんが，，，

そこで，スキューはいくつがいいと思いますか．また，最悪なくてもいいグラフがあったら知りたいです．

[このファイル](https://colab.research.google.com/drive/16GOsj8-g65sad4dKUUKZr4mYCGaJQxlo?usp=sharing)に出力が載っています．

それと，選択率ってなんですか？
選択率は二次元の場合だと，データ全体の面積から操作に使う面積の大きさを決定するパラメータだと思っていました．そして，その選ばれた面積の中で100万回検索するといった具合だと解釈しています．
桑村君に聞かれたときうまく答えられなかったので気になります．

**Comment by baycedar** [comment]
> スキューはいくつがいいと思いますか．また，最悪なくてもいいグラフがあったら知りたいです．

最少なら以下の4つで、書けるなら括弧の中含む8つだと思います。選択率は小さすぎるとほぼpoint lookup、大きすぎるとfull scanに近づいていくので、1E-3固定でグラフ数を減らしましょう。skew parameterは調整が難しいので0か0.5のいずれかでよいです。

次元横軸（縦軸は対数軸）
- skew: 0 (, 0.5)
- num. of records: 1E6
- selectivity: 1E-3

データ数横軸
- num. of dimensions: 2 (, 8)
- skew: 0
- selectivity: 1E-3

Skew横軸
- num. of records: 1E6
- num. of dimensions: 2 (, 8)
- selectivity: 1E-3

選択率横軸
- num. of records: 1E6
- num. of dimensions: 2 (, 8)
- skew: 0

> 選択率は二次元の場合だと，データ全体の面積から操作に使う面積の大きさを決定するパラメータだと思っていました．そして，その選ばれた面積の中で100万回検索するといった具合だと解釈しています．

データセット全体の最小包含矩形に対して範囲操作で使用する問合せ矩形が占める面積の割合であっています。ただし、各範囲操作命令は毎回別の問合せ矩形を使用します。

もう少し細かいことを言うと、問合せ矩形の各次元の範囲の長さはそれぞれ異なります（i.e., 超立方体ではない）。標準正規分布に従う乱数で各次元の重みを生成して、その重みに応じて各次元の範囲の大きさを決定しているので、ゆるい長方形のような形になっています。

**Comment by Mackyson** [comment]
> データセット全体の最小包含矩形に対して範囲操作で使用する問合せ矩形が占める面積の割合であっています。ただし、各範囲操作命令は毎回別の問合せ矩形を使用します。

選択率1%だとして，各範囲走査が毎回1%舐めるんでしょうか?
あるいは，すべての範囲走査が舐めた矩形のunionが1%になるのでしょうか?

**Comment by baycedar** [comment]
各範囲操作が1%です。
